### PR TITLE
[CPU] Optimize Broadcast node for case with scalar input

### PR DESCRIPTION
**Pull request by xuchen-intel**
### Details:
 - *Optimize Broadcast node for case with scalar input*

### Tickets:
 - *72654*

### CODE REVIEW

**Comment by xuchen-intel** [comment]
@dmitry-gorokhov Hi Dmitry, could you please take a look?

File: src/tests/functional/plugin/cpu/single_layer_tests/broadcast.cpp [hunk 1]
```diff
@@ -367,6 +427,19 @@ INSTANTIATE_TEST_CASE_P(smoke_DynamicShape5D, BroadcastLayerCPUTest,
                             ::testing::Values(CommonTestUtils::DEVICE_CPU)),
                         ::testing::Values(CPUSpecificParams{{}, {}, {}, "ref"})),
                     BroadcastLayerCPUTest::getTestCaseName);
+
+INSTANTIATE_TEST_CASE_P(smoke_DynamicShape5DScalar, BroadcastLayerCPUTest,
+                    ::testing::Combine(
+                        ::testing::Combine(
+                            ::testing::ValuesIn(dynamicInputShapesScalar),
+                            ::testing::Values(std::vector<int64_t>{8, 16, 1, 1, 7}),
```
Comment (dmitry-gorokhov) [inline src/tests/functional/plugin/cpu/single_layer_tests/broadcast.cpp:435]: Could you please also cover case when target_shape == {0} which is common case for dynamic models?

File: inference-engine/src/mkldnn_plugin/nodes/common/tile_broadcast_utils.cpp [hunk 2]
```diff
@@ -211,23 +211,62 @@ bool TileBroadcastCommon::prepareOptimizedParams(const MKLDNNNode *node, VectorD
...
-                                                 i4 * optimizedParams.srcStrides[4]);
-            auto dstData2 = dstData + (i0 * optimizedParams.dstStrides[0] + i1 * optimizedParams.dstStrides[1] +
-                                           i2 * optimizedParams.dstStrides[2] + i3 * optimizedParams.dstStrides[3] +
-                                           i4 * optimizedParams.dstStrides[4]);
-            for (int i = 0; i < optimizedParams.dims[5]; i++) {
-                cpu_memcpy(dstData2 + i * optimizedParams.dstStrides[5], srcData2, optimizedParams.dstStrides[5]);
+        if (optimizedParams.dstStrides[0] == optimizedParams.dims[5] * optimizedParams.dstStrides[5]) {
+            size_t data_size = optimizedParams.dstStrides[5];
+            size_t elt_cnt = optimizedParams.dims[5];
+            if (data_size == 1) {
```
Comment (dmitry-gorokhov) [inline inference-engine/src/mkldnn_plugin/nodes/common/tile_broadcast_utils.cpp:251]: This special case covers only char fill_values. In most cases the value has 4 bytes. Can we cover this as well? Smt like: 
```
auto srcData_i32 = reinterpret_cast<const int *>(srcMemory->GetPtr());
if (data_size == 4 && srcData_i32[0] == 0) {
  memset(dstData, 0, elt_cnt * data_size);
} else {
  broadcastScalar(srcData, dstData, elt_cnt, data_size);
}
```

File: inference-engine/src/mkldnn_plugin/nodes/common/tile_broadcast_utils.cpp [hunk 3]
```diff
@@ -211,23 +211,62 @@ bool TileBroadcastCommon::prepareOptimizedParams(const MKLDNNNode *node, VectorD
     return true;
 }
 
+void TileBroadcastCommon::broadcastScalar(const char *srcData, char *dstData, size_t elt_cnt, size_t data_size) {
```
Comment (dmitry-gorokhov) [inline inference-engine/src/mkldnn_plugin/nodes/common/tile_broadcast_utils.cpp:214]: The algo inside is a little bit unclear. Could you please add brief explanation?

File: src/tests/functional/plugin/cpu/single_layer_tests/broadcast.cpp (hunk 1 above)
Comment (xuchen-intel) [reply src/tests/functional/plugin/cpu/single_layer_tests/broadcast.cpp:435]: Fixed. Thanks Dmitry!

File: inference-engine/src/mkldnn_plugin/nodes/common/tile_broadcast_utils.cpp (hunk 2 above)
Comment (xuchen-intel) [reply inference-engine/src/mkldnn_plugin/nodes/common/tile_broadcast_utils.cpp]: Fixed. Thanks Dmitry!

File: inference-engine/src/mkldnn_plugin/nodes/common/tile_broadcast_utils.cpp (hunk 3 above)
Comment (xuchen-intel) [reply inference-engine/src/mkldnn_plugin/nodes/common/tile_broadcast_utils.cpp]: Fixed. Thanks Dmitry!

File: src/tests/functional/plugin/cpu/single_layer_tests/broadcast.cpp [hunk 4]
```diff
@@ -367,6 +427,44 @@ INSTANTIATE_TEST_CASE_P(smoke_DynamicShape5D, BroadcastLayerCPUTest,
...
+                {1}
+            }
+        }
+    }
+};
+
+INSTANTIATE_TEST_CASE_P(smoke_DynamicShapes1D, BroadcastLayerCPUTest,
+                    ::testing::Combine(::testing::Combine(
+                            ::testing::ValuesIn(dynamicShapes1D),
+                            ::testing::Values(std::vector<int64_t>{1}),
```
Comment (dmitry-gorokhov) [inline src/tests/functional/plugin/cpu/single_layer_tests/broadcast.cpp:460]: I mentioned {0} case not {1}.

Comment (xuchen-intel) [reply src/tests/functional/plugin/cpu/single_layer_tests/broadcast.cpp]: Fixed. Thanks Dmitry!

**Comment by xuchen-intel** [comment]
@dmitry-gorokhov CI is green now. Thanks Dmitry!

**Review by dmitry-gorokhov — APPROVED** [review]

**Pull request merged**